sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InsertMapSortInGroupingExpressions.scala

```diff
@@ -34,12 +35,46 @@ object InsertMapSortInGroupingExpressions extends Rule[LogicalPlan] {
     _.containsPattern(AGGREGATE), ruleId) {
     case a @ Aggregate(groupingExpr, _, _) =>
       val newGrouping = groupingExpr.map { expr =>
-        if (!expr.isInstanceOf[MapSort] && expr.dataType.isInstanceOf[MapType]) {
-          MapSort(expr)
+        if (!expr.exists(_.isInstanceOf[MapSort])
+          && expr.dataType.existsRecursively(_.isInstanceOf[MapType])) {
+          insertMapSortRecursively(expr)
         } else {
           expr
         }
       }
       a.copy(groupingExpressions = newGrouping)
   }
+
+  /*
+  Inserts MapSort recursively taking into account when
+  it is nested inside a struct or array.
+   */
+  private def insertMapSortRecursively(e: Expression): Expression = {
+    e.dataType match {
+      case _: MapType => MapSort(e)
```

Comment:
what if the key or value contains map type?

Reply:
So MapType as key type should be forbidden, but when MapType is type of value, this approach with MapSort won't work. We can create separate ticket for that one or try to address it as part of this PR? 
Most likely we would need to modify MapSort with that recursive approach, similar with what we have done for initial version of this PR.

Reply:
I think we can add map sort to the map value array and rebuild the map

Reply:
Hm, can you explain a bit, not sure what do you mean by add MapSort to value array since it works with expressions?

Reply:
map is basically two arrays for keys and values. We can use `MapKeys` and `MapValues` to get the two arrays, and `MapFromArrays` to rebuild the map. For arrays, we already know how to add `MapSort`

Reply:
Again, `CharVarcharUtils.stringLengthCheck` is a good example to follow

Reply:
Done and added corresponding tests.